File: 0213-house-robber-ii/0213-house-robber-ii.py
class Solution:
    # O(2n), O(n), backtracking + memoization
    def rob(self, nums: List[int]) -> int:
        @lru_cache(maxsize=None)
        def pick(start, end):
            if start <= end:
                money = nums[start]
                offset = 0
                if start == 0:
                    offset = 1

                return max(money + pick(start + 2, end - offset), pick(start + 1, end))
            else:
                return 0
        
        
        return pick(0, len(nums) - 1)

File: 0213-house-robber-ii/0213-house-robber-ii.cpp
/*
same as https://leetcode.com/problems/house-robber/
just do it once from 0 to n-1 and once from 1 to n and take max
*/
class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.size() == 1) return nums.front();
        return max(robrange(nums, 0, nums.size() - 1), robrange(nums, 1, nums.size()));
    }

private:
    int robrange(vector<int>& nums, int start, int end) {
        int iminus2 = 0, iminus1 = 0, i = 0;
        for (int x = start; x < end; ++x) {
            i = max(nums[x] + iminus2, iminus1);
            iminus2 = iminus1;
            iminus1 = i;
        }
        
        return i;
    }
};